miprometheus/problems/question_context_to_class/babiqa_dataset_single_question.py

```diff
@@ -0,0 +1,514 @@
+
+        if tenK:
+            self.dirname = os.path.join('tasks_1-20_v1-2', 'en-valid-10k')
+        else:
+            self.dirname = os.path.join('tasks_1-20_v1-2', 'en-valid')
+
+        if path is None:
+            path = self.download(root)
+
+        file_data = os.path.join(path, 'collected_' + data_type + outmod + '.txt')
+        with open(file_data, 'w') as tf:
+            for task in tasks:
+                with open(
+                        os.path.join(path,
+                                     'qa' + str(task) + '_' + data_type + '.txt')) as f:
+                    tf.write(f.read())
+        return self.parse(file_data, add_punctuation)
+
+    def download(self, root, check=None):
+
+        """Download and unzip an online archive (.zip, .gz, or .tgz).
+        Arguments:
+            root (str): Folder to download data to.
+            check (str or None): Folder whose existence indicates
+                that the dataset has already been downloaded, or
+                None to check the existence of root/{cls.name}.
+        Returns:
+            str: Path to extracted dataset.
+        """
+        #get path
+        path = os.path.join(root, self.name)
+        check = path if check is None else check
+
+        #download data
+        if not os.path.isdir(check):
+            for url in self.urls:
+                if isinstance(url, tuple):
+                    url, filename = url
+                else:
+                    filename = os.path.basename(url)
+                zpath = os.path.join(path, filename)
+                if not os.path.isfile(zpath):
+                    if not os.path.exists(os.path.dirname(zpath)):
+                        os.makedirs(os.path.dirname(zpath))
+                    print('downloading {}'.format(filename))
+                    self.download_from_url(url, zpath)
+                zroot, ext = os.path.splitext(zpath)
+                _, ext_inner = os.path.splitext(zroot)
+
+                #unzip the data
+                if ext == '.zip':
+                    with zipfile.ZipFile(zpath, 'r') as zfile:
+                        print('extracting')
+                        zfile.extractall(path)
+
+                # tarfile cannot handle bare .gz files
+                elif ext == '.tgz' or ext == '.gz' and ext_inner == '.tar':
+                    with tarfile.open(zpath, 'r:gz') as tar:
+                        dirs = [member for member in tar.getmembers()]
+                        tar.extractall(path=path, members=dirs)
+
+                #in case it is a gz file
+                elif ext == '.gz':
+                    with gzip.open(zpath, 'rb') as gz:
+                        with open(zroot, 'wb') as uncompressed:
+                            shutil.copyfileobj(gz, uncompressed)
+
+        #Return path to extracted dataset
+        return os.path.join(path, self.dirname)
```

Comment:
This method looks quite generic, which is nice, so I would move it up in the classes hierarchy

Reply:
ok